src/compile-gate.lisp

```diff
@@ -155,8 +221,113 @@ QUBITS should be a NAT-TUPLE of qubits representing the Hilbert space."
                  operator
                  amps
                  arefs))))
-           :dotimes-iterator 'lparallel:pdotimes)))))
+           :dotimes-iterator +dotimes-iterator+)))))
 
+(defun generate-gate-application-code-with-matrix (qubits matrix)
+  "Generate a lambda form which takes two arguments, a QUANTUM-OPERATOR and a QUANTUM-STATE, which efficiently applies that operator to that state, lifted from the Hilbert space designated by QUBITS.
+
+QUBITS should be a NAT-TUPLE of qubits representing the Hilbert space."
+  (check-type qubits nat-tuple)
+  (let* ((num-gate-qubits (nat-tuple-cardinality qubits))
+         (operator-size (expt 2 num-gate-qubits)))
+    (alexandria:with-gensyms (wavefunction)
+      `(lambda (,wavefunction)
+         (declare ,*optimize-dangerously-fast*
+                  (type quantum-state ,wavefunction))
+         ,(generate-complement-iteration
+           qubits
+           wavefunction
+           (lambda (addr)
+             (generate-extraction-code
+              addr
+              wavefunction
+              qubits
+              (lambda (amps arefs)
+                (generate-inner-matrix-multiply-code
+                 operator-size
+                 matrix
+                 amps
+                 arefs))))
+           :dotimes-iterator +dotimes-iterator+)))))
+
+(defvar *unrolling-size* 0 "Maximum number of loop body duplications in loop unrolling. Zero disables unrolling.")
```

Comment:
this + generate-...-measurement added in https://github.com/rigetti/qvm/pull/130